src/charm.py

```diff
@@ -7,29 +7,176 @@
+            self._prometheus_source.on.sources_changed,  # pyright: ignore
+            self.reconcile,
+        )
+        # Not sure we need this, but kept it here for completeness.
+        self.framework.observe(
+            self._prometheus_source.on.sources_to_delete_changed,  # pyright: ignore
+            self.reconcile,
+        )
+
+        self.framework.observe(self.on.collect_unit_status, self.on_collect_status)
+        self.framework.observe(self.on.config_changed, self.reconcile)
+
+        # Expose the Kiali workload through the service
+        self.unit.set_ports(Port("tcp", KIALI_PORT))
+
+    def on_collect_status(self, e: ops.CollectStatusEvent):
+        """Collect and report the statuses of the charm."""
+        # TODO: Extract the generation of these statuses to a helper method, that way other parts of the charm can know
+        #  if the charm is active too.
+        statuses: List[StatusBase] = []
+
+        if not self._container.can_connect():
+            statuses.append(ops.WaitingStatus("Waiting for the Kiali container to be ready"))
+
+        if not (prometheus_source_available := self._is_prometheus_source_available()):
+            statuses.append(ops.BlockedStatus("Prometheus source is not available"))
+
+        if prometheus_source_available:
+            # Only valid if we have a prometheus source
+            kiali_config = self._generate_kiali_config()
+            kiali_local_url = f"http://localhost:{kiali_config['server']['port']}/kiali"
+            if not _is_kiali_available(kiali_local_url):
+                statuses.append(
+                    ops.WaitingStatus(
+                        "Kiali is configured and container is ready, but Kiali is not available"
+                    )
+                )
+
+        if len(statuses) == 0:
+            statuses.append(ops.ActiveStatus())
+
+        for status in statuses:
+            e.add_status(status)
+
+    def reconcile(self, _event: ops.ConfigChangedEvent):
         """Reconcile the entire state of the charm."""
-        self.unit.status = ops.BlockedStatus("todo: implement reconcile")
+        self._configure_kiali_workload()
+
+    def _configure_kiali_workload(self):
+        """Configure the Kiali workload, if possible, logging errors otherwise.
+
+        This will generate and push the Kiali configuration to the container, restarting the service if necessary.
+        The purpose here is that this should always attempt to configure/start Kiali, but it does not guarantee Kiali is
+        running after completion.  If any known errors occur, they will be logged and this method will return without
+        error.  To confirm if Kiali is working, check the status of the Kiali workload directly.
+        """
+        # TODO: Can we make this generic and share with other charms?
+        name = "kiali"
+        if not self._container.can_connect():
+            LOGGER.warning(f"Container is not ready, cannot configure {name}")
```

Comment:
I do not think we need a warning here. This is not really an indication that anything is wrong. Changing to debug would be appropriate I think.

Reply:
Agreed, changed to debug.  I could see it being `info` level too, but I'm good either way